### PR TITLE
[Core][Proposal] GetSpecifications method on Elements/Conditions

### LINKED ISSUE #5093: [Discussion][Proposal] Requirements method on Elements/Conditions
After speaking with @rubenzorrilla I thought on a method in order to automatically add variable/dofs/etc... This method would be complementary to Check method. My idea is a method like:

~~~cpp
Parameters GetRequirements(const ProcessInfo& rProcessInfo) const override
{
	const Parameters requirements = Parameters("{
		"required_variables" : ["DISPLACEMENT_X", "DISPLACEMENT_Y", "DISPLACEMENT_Z"],
		"required_dofs"      : ["DISPLACEMENT_X", "DISPLACEMENT_Y", "DISPLACEMENT_Z"]
	}");
	return requirements;
}
~~~

This method can be called for example by the check method, but additionally the idea of the proposal is to be called when reading a mdpa file or in the solver, and add the missing dofs/variables/properties and everything we can think of. The good thing of using Parameters on this method is than in the future we can extend it to any need, the bad thing is that Parameters is a heavy object, but if we do only once per element is not so bad.

The origin of this was the need to put on the Parameters "requires_rotation_dofs" when using shells or beams.

### PR DESCRIPTION
**Pull request by loumalouomega**
Fixes #5093, as discussed in #5093

### CODE REVIEW

**Comment by AlejandroCornejo** [comment]
The @KratosMultiphysics/implementation-committee has reached consensus with this.

**Comment by pooyan-dadvand** [comment]
What I didn't deduce in the previous discusion, why we need to pass the process info?

**Comment by loumalouomega** [comment]
> What I didn't deduce in the previous discusion, why we need to pass the process info?

In case some additional information may be needed. It is my proposal, i can remove it

**Comment by philbucher** [comment]
I would leave the ProcessInfo, it is used in pretty much all methods

What I don't like is adding the `Kratos::Parameters` as include, this IMO introduces quite a large dependency (and we already have quite strong dependencies)

Maybe a solution would be to separate into header/source (@loumalouomega please wait for others opinions to avoid overhead)

**Comment by loumalouomega** [comment]
> I would leave the ProcessInfo, it is used in pretty much all methods
> 
> What I don't like is adding the `Kratos::Parameters` as include, this IMO introduces quite a large dependency (and we already have quite strong dependencies)
> 
> Maybe a solution would be to separate into header/source (@loumalouomega please wait for others opinions to avoid overhead)

The problem with that is that you require

**Comment by loumalouomega** [comment]
> I would leave the ProcessInfo, it is used in pretty much all methods
> 
> What I don't like is adding the `Kratos::Parameters` as include, this IMO introduces quite a large dependency (and we already have quite strong dependencies)
> 
> Maybe a solution would be to separate into header/source (@loumalouomega please wait for others opinions to avoid overhead)

I assume you mean forward declaration?, in that case you need to work with references or raw pointers (it is the only way for the compiler to assign a memory without knowing the object)

**Comment by philbucher** [comment]
ah sure, didn't think this through...

no more complaints from me then :)

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee we agree about API without passing the ProcessInfo as follow:

```c++
const Parameters GetSpecifications() const;
```

And we also discussed **without achieving the final design** about the contents of this info for the elements and conditions basing on the XML files we provide for GUI:

```Json
{
        "framework" : "eulerian",
        "symmetric_lhs": true,
        "positive_definite_lhs": false,
	"required_variables": ["DISPLACEMENT"],
	"required_dofs": ["DISPLACEMENT_X", "DISPLACEMENT_Y", "DISPLACEMENT_Z"],
        "flags_used": ["BOUNDARY", "ACTIVE"],
        "compatible_geometries": ["Tetrahedra3D4", "Hexahedra3D8"],
        "compatible_constitutive_laws": {
            "type" : ["PlaneStress","PlaneStrain"],
            "dimension" : "2D",
            "strain_size" : "3"
        }
}
```

This can be used as a basis for the discussion. Please give us your feedback to improve this proposal.

**Comment by roigcarlo** [comment]
@rubenzorrilla suggests adding some information about the time integration in the list above. 
```json
    `is_implicit`: true,
    `element_integrates_in_time`: true
```

**Comment by loumalouomega** [comment]
> 
> 
> @rubenzorrilla suggests adding some information about the time integration in the list above.
> 
> ```json
>     `is_implicit`: true,
>     `element_integrates_in_time`: true
> ```

I thinks as far this PR just adds the Parameters, the content can be discussed later

**Comment by roigcarlo** [comment]
> > @rubenzorrilla suggests adding some information about the time integration in the list above.
> > ```json
> >     `is_implicit`: true,
> >     `element_integrates_in_time`: true
> > ```
> 
> I thinks as far this PR just adds the Parameters, the content can be discussed later

@KratosMultiphysics/technical-committee prefers to discuss the contents together with the API to avoid future misuses of the new method.

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee  also asks for a test to check that all the elements registered in Kratos comply with the standard fields of this specification:
* Please propose the first version of this specification in this PR taking into account the suggestion above.
* Test should loop over all registered elements:  if the specification is not empty, check if all fields are matched with standard fields. if is empty is just bypass it to maintain the backward compatibility.

**Comment by loumalouomega** [comment]
> 
> 
> @KratosMultiphysics/technical-committee also asks for a test to check that all the elements registered in Kratos comply with the standard fields of this specification:
> 
>     * Please propose the first version of this specification in this PR taking into account the suggestion above.

This means to add the test in this PR or just to propose the test structure? This will depend on the physical problem?. In that case may be added as a static method in the application definition. Or the previously stated is the minimum that should be considered in every problem?
> 
>     * Test should loop over all registered elements:  if the specification is not empty, check if all fields are matched with standard fields. if is empty is just bypass it to maintain the backward compatibility.

OK, see my previous comment.

**Comment by pooyan-dadvand** [comment]
Both in this PR. The idea is to discuss the fields before merge to impose the coherency from beginning.

**Comment by adityaghantasala** [comment]
On behalf of @KratosMultiphysics/implementation-committee  

We were thinking it would be benificial to have some empty set of keys in the base element/condition. Current implementation passes  empty Parameters. But having a set of consentuated set of keys like `required_variables` and `require_dofs` (every element/condition requires dofs and variables) etc would encourage developers to think of kratos element/condition in a more specific way.  Having an empty parameters object there may encourage misuse of this nice functionality. 

@loumalouomega  could you already propose the first version of this set ?

**Comment by loumalouomega** [comment]
> @loumalouomega could you already propose the first version of this set ?

I need to implement the test. but it is almost ready

**Comment by loumalouomega** [comment]
@KratosMultiphysics/implementation-committee Tests have been added

**Comment by RiccardoRossi** [comment]
some minor comments after reviewing in @KratosMultiphysics/technical-committee :

1 - is_implicit -> time_integration : [ ... here list possible compatible types ... "implicit", "explicit", "static"]
2 - output --> missing to list potential outputs. so 
    - output on Gauss Points
    - nodal historical output
    - nodal non historical output
    - entity output (the result of GetValue on the element)
3 - we should define a map geometry_type/name
4 - more filter features for CL, in particular deformation measure and stress measure required
5 - "Documentation" add a free field where one can point to the documentation and maybe a reference to the formulation

The idea is to use this as a mechanism for generating automatic docs that go along with the code. For this to work we will need to use this extensively in the testing (which will have to be done app by app to take into account the specificities of each)

**Comment by loumalouomega** [comment]
> some minor comments after reviewing in @KratosMultiphysics/technical-committee :
> 
> 1 - is_implicit -> time_integration : [ ... here list possible compatible types ... "implicit", "explicit", "static"]

Static is implicit... (I think we are mixing time integration with resolution type)

> 2 - output --> missing to list potential outputs. so
> - output on Gauss Points
> - nodal historical output
> - nodal non historical output
> - entity output (the result of GetValue on the element)

OK

> 3 - we should define a map geometry_type/name

??¿¿

> 4 - more filter features for CL, in particular deformation measure and stress measure required

I require a defined list in order to implement it correctly

> 5 - "Documentation" add a free field where one can point to the documentation and maybe a reference to the formulation

OK

> 
> The idea is to use this as a mechanism for generating automatic docs that go along with the code. For this to work we will need to use this extensively in the testing (which will have to be done app by app to take into account the specificities of each)

OK

File: kratos/utilities/specifications_utilities.h [hunk 1]
```diff
@@ -52,6 +52,26 @@ namespace Kratos
...
+        Explicit = 2
+    };
+
+    /**
+     * @brief This enum defines a "hash" used to identify if Lagrangian/Eulerian or ALE framework is considered
+     */
+    enum class Framework
+    {
+        Lagrangian = 0,
+        Euler      = 1,
```
Comment (rubenzorrilla) [inline kratos/utilities/specifications_utilities.h]:
```suggestion
        Eulerian      = 1,
```

File: kratos/utilities/specifications_utilities.h [hunk 2]
```diff
@@ -52,6 +52,26 @@ namespace Kratos
...
+        ALE        = 2
+    };
+
```
Comment (rubenzorrilla) [inline kratos/utilities/specifications_utilities.h:74]: Should we add the Updated Lagrangian to the list too?

File: kratos/utilities/specifications_utilities.cpp [hunk 3]
```diff
@@ -442,15 +552,15 @@ std::string DetermineFramework(ModelPart& rModelPart)
                 framework = r_framework;
             } else {
                 CompareElementsAndConditionsUtility::GetRegisteredName(*it_cond_begin, condition_name);
-                KRATOS_WARNING("SpecificationsUtilities") << "Framework:" << r_framework << " is used by the condition: " << condition_name << ". But also the framework: " << framework << " is used" << std::endl; 
+                KRATOS_WARNING("SpecificationsUtilities") << "Framework:" << r_framework << " is used by the condition: " << condition_name << ". But also the framework: " << framework << " is used" << std::endl;
```
Comment (rubenzorrilla) [inline kratos/utilities/specifications_utilities.cpp:523]: Not sure if this has to be a warning. It is perfectly possible to mix Eulerian elements with ALE ones.

File: kratos/utilities/specifications_utilities.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/utilities/specifications_utilities.h:74]: Nope, Updated Lagrangian is still Lagrangian. The point of Lagrangian is communicate to the solver that the nodes of these elements/conditions require to be moved after solving the step

Comment (loumalouomega) [reply kratos/utilities/specifications_utilities.h:74]: Well, fort now this just tells the framework, for now it is not taken into account for nothing

File: kratos/utilities/specifications_utilities.cpp (hunk 3 above)
Comment (loumalouomega) [reply kratos/utilities/specifications_utilities.cpp:523]: And Lagrangian with Eulerian, or ALE, I agree. This is a draft, I am doing what the @KratosMultiphysics/technical-committee  tells me, and the lack of details makes me to take some ambiguous decisions

File: kratos/utilities/specifications_utilities.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/utilities/specifications_utilities.h]: BTW, for now this is not used. Currently I use the string, this is just an idea to later remove if the @KratosMultiphysics/technical-committee says so. PS: enums are a pain in python

File: kratos/utilities/specifications_utilities.cpp (hunk 3 above)
Comment (philbucher) [reply kratos/utilities/specifications_utilities.cpp:523]: just to say that we are also not sure what should be there...

Comment (loumalouomega) [reply kratos/utilities/specifications_utilities.cpp:523]: > just to say that we are also not sure what should be there...

imagen

File: kratos/utilities/specifications_utilities.h (hunk 2 above)
Comment (rubenzorrilla) [reply kratos/utilities/specifications_utilities.h:74]: Your argument makes sense to me (the updated is still Lagrangian).

**Comment by rubenzorrilla** [comment]
@KratosMultiphysics/implementation-committee which is the status of this?

**Comment by loumalouomega** [comment]
> @KratosMultiphysics/implementation-committee which is the status of this?

This is done with the current specifications

File: applications/StructuralMechanicsApplication/custom_conditions/base_load_condition.cpp [hunk 4]
```diff
@@ -366,6 +366,38 @@ void BaseLoadCondition::AddExplicitContribution(
...
+            "gauss_point"            : [],
+            "nodal_historical"       : [],
+            "nodal_non_historical"   : [],
+            "entity"                 : []
+        },
+        "required_variables"         : ["DISPLACEMENT"],
+        "required_dofs"              : ["DISPLACEMENT_X","DISPLACEMENT_Y","DISPLACEMENT_Z"],
+        "flags_used"                 : [],
+        "compatible_geometries"      : ["Line2D2", "Line2D3", "Triangle3D3", "Triangle3D6", "Quadrilateral3D4", "Quadrilateral3D8", "Quadrilateral3D9"],
+        "is_implicit"                : true,
```
Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_conditions/base_load_condition.cpp]: It is redundant as it is given in the "time_integration" field.

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_conditions/base_load_condition.cpp]: I just follow the requirements from the @KratosMultiphysics/technical-committee, if the @KratosMultiphysics/technical-committee agrees I will remove it

Comment (pooyan-dadvand) [reply applications/StructuralMechanicsApplication/custom_conditions/base_load_condition.cpp]: @KratosMultiphysics/technical-committee agrees on removing it and check the `is_implicit `by `[time_integration].Has("implicit")`

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee agrees on the addition of the `GetSpecification `method and its current content (removing the is_implicit) 

However, an important concern is to avoid slow performance of this mechanism in critical locations. This means that this method should be used ONLY in the input or other places which are done once and the overhead is assumable. 

Please add this performance comment to the doxygen of the method so it is presented to all developers.

**Comment by loumalouomega** [comment]
> @KratosMultiphysics/technical-committee agrees on the addition of the `GetSpecification `method and its current content (removing the is_implicit)
> 
> However, an important concern is to avoid slow performance of this mechanism in critical locations. This means that this method should be used ONLY in the input or other places which are done once and the overhead is assumable.
> 
> Please add this performance comment to the doxygen of the method so it is presented to all developers.

Done

**Comment by tteschemacher** [comment]
I know I'm late, but do you think we could also add required polynomial degree of geometry?

**Comment by loumalouomega** [comment]
> I know I'm late, but do you think we could also add required polynomial degree of geometry?

If @KratosMultiphysics/technical-committee agrees...

**Comment by rubenzorrilla** [comment]
> > I know I'm late, but do you think we could also add required polynomial degree of geometry?
> 
> If @KratosMultiphysics/technical-committee agrees...

IMO this `required_polynomial_degree_of_geometry` is way specific of IGA formulations to be included in the standard specifications.

**Comment by tteschemacher** [comment]
> > > I know I'm late, but do you think we could also add required polynomial degree of geometry?
> > 
> > 
> > If @KratosMultiphysics/technical-committee agrees...
> 
> IMO this `required_polynomial_degree_of_geometry` is way specific of IGA formulations to be included in the standard specifications.

But we're comparing and updating towards defaults, no? The default can be 1. Nevertheless, I know that there are elements in the structural mechanics application which work with a higher polynomial degree. E.g. the beam.

**Comment by pooyan-dadvand** [comment]
I think is not a bad idea. It can also be applied to the normal geometries.

**Comment by loumalouomega** [comment]
> I think is not a bad idea. It can also be applied to the normal geometries.

Then, do I add it?

**Comment by loumalouomega** [comment]
> > I think is not a bad idea. It can also be applied to the normal geometries.
> 
> Then, do I add it?

What do I do?

**Comment by loumalouomega** [comment]
> > > I think is not a bad idea. It can also be applied to the normal geometries.
> > 
> > 
> > Then, do I add it?
> 
> What do I do?

Ping @KratosMultiphysics/technical-committee

**Comment by pooyan-dadvand** [comment]
We agree with adding the `required_polynomial_degree_of_geometry`

**Comment by loumalouomega** [comment]
> We agree with adding the `required_polynomial_degree_of_geometry`

It is a bool? @tteschemacher

**Comment by rubenzorrilla** [comment]
> > We agree with adding the `required_polynomial_degree_of_geometry`
> 
> It is a bool? @tteschemacher

I think it's a positive integer.

**Comment by loumalouomega** [comment]
> > > We agree with adding the `required_polynomial_degree_of_geometry`
> > 
> > 
> > It is a bool? @tteschemacher
> 
> I think it's a positive integer.

I will remove required_ then, is confusing. And it will be an integer, because the json does not support unsigned integers

**Comment by tteschemacher** [comment]
> > > > We agree with adding the `required_polynomial_degree_of_geometry`
> > > 
> > > 
> > > It is a bool? @tteschemacher
> > 
> > 
> > I think it's a positive integer.
> 
> I will remove required_ then, is confusing. And it will be an integer, because the json does not support unsigned integers

Thanks Vicente, yes it should be an int. Maybe we say minimal_required_p...?

**Comment by loumalouomega** [comment]
> `required_polynomial_degree_of_geometry`

Done

**Comment by loumalouomega** [comment]
@KratosMultiphysics/technical-committee this is ready

**Comment by loumalouomega** [comment]
> @KratosMultiphysics/technical-committee this is ready

Ping

**Review by pooyan-dadvand — APPROVED** [review]
Thanks @loumalouomega for the effort!

**Pull request merged**